ArmPkg/Drivers/ArmPsciMpServicesDxe/ArmPsciMpServicesDxe.c

```diff
@@ -0,0 +1,1767 @@
+                               BSP. The range is from 0 to the total number of
+                               logical processors minus 1. The total number of
+                               logical processors can be retrieved by
+                               EFI_MP_SERVICES_PROTOCOL.GetNumberOfProcessors().
+
+  @retval EFI_SUCCESS             The current processor handle number was returned
+                                  in ProcessorNumber.
+  @retval EFI_INVALID_PARAMETER   ProcessorNumber is NULL.
+
+**/
+STATIC
+EFI_STATUS
+EFIAPI
+WhoAmI (
+  IN EFI_MP_SERVICES_PROTOCOL  *This,
+  OUT UINTN                    *ProcessorNumber
+  )
+{
+  UINTN   Index;
+  UINT64  ProcessorId;
+
+  if (ProcessorNumber == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  ProcessorId = GET_MPIDR_AFFINITY_BITS (ArmReadMpidr ());
+  for (Index = 0; Index < mCpuMpData.NumberOfProcessors; Index++) {
+    if (ProcessorId == gProcessorIDs[Index]) {
+      *ProcessorNumber = Index;
+      break;
+    }
+  }
+
+  return EFI_SUCCESS;
+}
+
+STATIC EFI_MP_SERVICES_PROTOCOL  mMpServicesProtocol = {
+  GetNumberOfProcessors,
+  GetProcessorInfo,
+  StartupAllAPs,
+  StartupThisAP,
+  SwitchBSP,
+  EnableDisableAP,
+  WhoAmI
+};
+
+/** Adds the specified processor the list of failed processors.
+
+   @param ProcessorIndex The processor index to add.
+   @param ApState        Processor state.
+
+**/
+STATIC
+VOID
+AddProcessorToFailedList (
+  UINTN      ProcessorIndex,
+  CPU_STATE  ApState
+  )
+{
+  UINTN    Index;
+  BOOLEAN  Found;
+
+  Found = FALSE;
+
+  if (ApState == CpuStateIdle) {
+    return;
+  }
+
+  // If we are retrying make sure we don't double count
+  for (Index = 0; Index < mCpuMpData.FailedListIndex; Index++) {
+    if (mCpuMpData.FailedList[Index] == ProcessorIndex) {
+      Found = TRUE;
+      break;
+    }
+  }
+
+  /* If the CPU isn't already in the FailedList, add it */
+  if (!Found) {
+    mCpuMpData.FailedList[mCpuMpData.FailedListIndex++] = Index;
+  }
+}
+
+/** Handles the StartupAllAPs case where the timeout has occurred.
+
+**/
+STATIC
+VOID
+ProcessStartupAllAPsTimeout (
+  VOID
+  )
+{
+  CPU_AP_DATA  *CpuData;
+  UINTN        Index;
+
+  if (mCpuMpData.FailedList == NULL) {
+    return;
+  }
+
+  for (Index = 0; Index < mCpuMpData.NumberOfProcessors; Index++) {
+    CpuData = &mCpuMpData.CpuData[Index];
+    if (IsProcessorBSP (Index)) {
+      // Skip BSP
+      continue;
+    }
+
+    if (!IsProcessorEnabled (Index)) {
+      // Skip Disabled processors
+      continue;
+    }
+
+    CpuData = &mCpuMpData.CpuData[Index];
+    AddProcessorToFailedList (Index, GetApState (CpuData));
+  }
+}
+
+/** Updates the status of the APs.
+
+   @param[in] ProcessorIndex The index of the AP to update.
+**/
+STATIC
+VOID
+UpdateApStatus (
+  IN UINTN  ProcessorIndex
+  )
+{
+  EFI_STATUS   Status;
+  CPU_AP_DATA  *CpuData;
+  CPU_AP_DATA  *NextCpuData;
+  CPU_STATE    State;
+  UINTN        NextNumber;
+

```

Comment:
Check that ProcessorIndex doesn't exceed the total number of processors available?

Reply:
Thanks, updated in the latest push.